packages/inferno-redux/src/components/connectAdvanced.ts

```diff
@@ -0,0 +1,320 @@
+				// componentWillUnmount do not. Because of this, trySubscribe happens during ...didMount.
+				// Otherwise, unsubscription would never take place during SSR, causing a memory leak.
+				// To handle the case where a child component may have triggered a state change by
+				// dispatching an action in its componentWillMount, we have to re-run the select and maybe
+				// re-render.
+				this.subscription.trySubscribe();
+				this.selector.run(this.props);
+				if (this.selector.shouldComponentUpdate) {
+					this.forceUpdate();
+				}
+			}
+
+			componentWillReceiveProps(nextProps) {
+				this.selector.run(nextProps);
+			}
+
+			shouldComponentUpdate() {
+				return this.selector.shouldComponentUpdate;
+			}
+
+			componentWillUnmount() {
+				if (this.subscription) {
+					this.subscription.tryUnsubscribe();
+				}
+
+				// these are just to guard against extra memory leakage if a parent element doesn't
+				// dereference this instance properly, such as an async callback that never finishes
+				this.subscription = null;
+				this.store = null;
+				this.parentSub = null;
+				this.selector.run = () => {};
+			}
+
+			getWrappedInstance() {
+				invariant(
+					withRef,
+					`To access the wrapped instance, you need to specify ` +
+					`{ withRef: true } in the options argument of the ${methodName}() call.`,
+				);
+
+				return this.wrappedInstance;
+			}
+
+			private setWrappedInstance(ref) {
+				this.wrappedInstance = ref;
+			}
+
+			initSelector() {
+				const { dispatch } = this.store;
+				const { getState } = this;
+				const sourceSelector = selectorFactory(dispatch, selectorFactoryOptions);
+				const runComponentSelector = (props: any) => {
+					try {
+							const nextProps = sourceSelector(getState(), props);
+							if (selector.error || nextProps !== selector.props) {
+								selector.shouldComponentUpdate = true;
+								selector.props = nextProps;
+								selector.error = null;
+							}
+						} catch (error) {
+							selector.shouldComponentUpdate = true;
+							selector.error = error;
+						}
+				};
+
+				// wrap the selector in an object that tracks its results between runs
+				this.selector = {
+					shouldComponentUpdate: true,
+					props: sourceSelector(getState(), this.props),
+					run: runComponentSelector,
+					error: null,
+				};
+				const selector = this.selector;
+			}
+
+			initSubscription() {
+				if (shouldHandleStateChanges) {
+					const subscription = this.subscription = new Subscription(this.store, this.parentSub);
+					const dummyState = {};
+
+					const onStateChange = () => {
+						this.selector.run(this.props);
+
+						if (!this.selector.shouldComponentUpdate) {
+							subscription.notifyNestedSubs();
+						} else {
+							this.componentDidUpdate = () => {
+								this.componentDidUpdate = undefined;
+								subscription.notifyNestedSubs();
+							};
+
+							this.setState(dummyState);
```

Comment:
Most of the tests rely on atomic updates. That is, after doing a `dispatch`, I test that I get the correct props/state from the components. This mirros how `react-redux` does thing (and in fact it's just a port of the code as is). Since asynchronity can break test behaviour, it stands to reason it can break actual behaviour as well.